[JSC] Implement atomic operations in new wasm baseline JIT

Constellation opened the pull request:
#### 30b2907267b216913a9e14bfa65ab27d6b1a5b4b
<pre>
[JSC] Implement atomic operations in new wasm baseline JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=252457">https://bugs.webkit.org/show_bug.cgi?id=252457</a>
rdar://105581013

Reviewed by Justin Michaud.

This patch implements atomics in new wasm baseline JIT. And it also fixes some issues.

1. Implement all wasm atomics in new wasm baseline JIT.
2. We found that new baseline&apos;s emitCCall&apos;s results are wrong since C-Call and Wasm-Call have difference in calling convention,
   in particular, return values. We, for now, fix the return values. We should decouple Wasm-Call calling convention things from
   emitCCall since this is C-Call.
3. Fix disassembly conditions.
4. Show AssemblyComments when needDisassemblySupport is true.

* Source/JavaScriptCore/assembler/AssemblyComments.h:
(JSC::AssemblyCommentRegistry::registerCodeRange):
(JSC::AssemblyCommentRegistry::unregisterCodeRange):
(JSC::AssemblyCommentRegistry::comment):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::sizeOfLoadOp):
(JSC::Wasm::BBQJIT::typeOfLoadOp):
(JSC::Wasm::BBQJIT::accessWidth):
(JSC::Wasm::BBQJIT::sizeOfAtomicOpMemoryAccess):
(JSC::Wasm::BBQJIT::emitSanitizeAtomicResult):
(JSC::Wasm::BBQJIT::emitAtomicOpGeneric):
(JSC::Wasm::BBQJIT::emitAtomicLoadOp):
(JSC::Wasm::BBQJIT::atomicLoad):
(JSC::Wasm::BBQJIT::emitAtomicStoreOp):
(JSC::Wasm::BBQJIT::atomicStore):
(JSC::Wasm::BBQJIT::emitAtomicBinaryRMWOp):
(JSC::Wasm::BBQJIT::atomicBinaryRMW):
(JSC::Wasm::BBQJIT::emitAtomicCompareExchange):
(JSC::Wasm::BBQJIT::atomicCompareExchange):
(JSC::Wasm::BBQJIT::atomicWait):
(JSC::Wasm::BBQJIT::atomicNotify):
(JSC::Wasm::BBQJIT::atomicFence):
(JSC::Wasm::BBQJIT::emitCCall):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::didCompleteCompilation):

Canonical link: <a href="https://commits.webkit.org/260460@main">https://commits.webkit.org/260460@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0886a8b80b4f463f0e1ea308b01f3cf662bf2ed2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108282 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17373 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117400 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116753 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8652 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100497 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114051 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97339 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42054 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96074 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28983 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83736 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97477 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10214 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30327 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96826 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8327 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10953 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7230 "Passed tests") | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16363 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49924 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105865 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12539 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26206 "Passed tests") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3929 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/assembler/AssemblyComments.h:
```diff
@@ -88,7 +88,7 @@ class AssemblyCommentRegistry {
     inline std::optional<String> comment(void* in)
     {
-        if (LIKELY(!Options::dumpDisassembly()))
+        if (LIKELY(!Options::needDisassemblySupport()))
```
Nice

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5265,13 +6248,50 @@ class BBQJIT {
+            // exclude all possible return value registers from ScratchScope so we can guarantee there's never any interference.
+            currentBinding = RegisterBinding::none();
+            if (resultLocation.isGPR())
+                m_gprSet.add(resultLocation.asGPR(), Width::Width64);
```
IgnoreVectors

Constellation replied:
Changed.

Constellation replied:
Ah, actually, all `m_gprSet` uses Width64. So in this patch, I don't change it.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5265,13 +6248,50 @@ class BBQJIT {
-        // TODO: Handle returning multiple values from a C call
-        ASSERT(returnedValues.size() == 1);
-        result = returnedValues[0];
+        RegisterBinding& currentBinding = resultLocation.isGPR() ? m_gprBindings[resultLocation.asGPR()] : m_fprBindings[resultLocation.asFPR()];
```
Can we remove all bindings that might ever interfere, that way any bugs will be easier to catch? Maybe we can move this into a helper and do it anywhere that a scratch might get clobbered?

Constellation replied:
saveValuesAcrossCall is doing it.

justinmichaud reviewed (approved):
r=me

Constellation committed:
[JSC] Implement atomic operations in new wasm baseline JIT
https://bugs.webkit.org/show_bug.cgi?id=252457
rdar://105581013

Reviewed by Justin Michaud.

This patch implements atomics in new wasm baseline JIT. And it also fixes some issues.

1. Implement all wasm atomics in new wasm baseline JIT.
2. We found that new baseline's emitCCall's results are wrong since C-Call and Wasm-Call have difference in calling convention,
   in particular, return values. We, for now, fix the return values. We should decouple Wasm-Call calling convention things from
   emitCCall since this is C-Call.
3. Fix disassembly conditions.
4. Show AssemblyComments when needDisassemblySupport is true.

* Source/JavaScriptCore/assembler/AssemblyComments.h:
(JSC::AssemblyCommentRegistry::registerCodeRange):
(JSC::AssemblyCommentRegistry::unregisterCodeRange):
(JSC::AssemblyCommentRegistry::comment):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::sizeOfLoadOp):
(JSC::Wasm::BBQJIT::typeOfLoadOp):
(JSC::Wasm::BBQJIT::accessWidth):
(JSC::Wasm::BBQJIT::sizeOfAtomicOpMemoryAccess):
(JSC::Wasm::BBQJIT::emitSanitizeAtomicResult):
(JSC::Wasm::BBQJIT::emitAtomicOpGeneric):
(JSC::Wasm::BBQJIT::emitAtomicLoadOp):
(JSC::Wasm::BBQJIT::atomicLoad):
(JSC::Wasm::BBQJIT::emitAtomicStoreOp):
(JSC::Wasm::BBQJIT::atomicStore):
(JSC::Wasm::BBQJIT::emitAtomicBinaryRMWOp):
(JSC::Wasm::BBQJIT::atomicBinaryRMW):
(JSC::Wasm::BBQJIT::emitAtomicCompareExchange):
(JSC::Wasm::BBQJIT::atomicCompareExchange):
(JSC::Wasm::BBQJIT::atomicWait):
(JSC::Wasm::BBQJIT::atomicNotify):
(JSC::Wasm::BBQJIT::atomicFence):
(JSC::Wasm::BBQJIT::emitCCall):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::didCompleteCompilation):

Canonical link: https://commits.webkit.org/260460@main
